### PR TITLE
use HttpRequestMock to mock Luis

### PR DESCRIPTION
**Pull request by xieofxie**
Part of #4451 

## Description

Use HttpRequestMock to mock Luis.
Fix IncludeAPIResults not working in LuisAdaptiveRecognizer

## Specific Changes

  - Add Content to HttpRequestSequenceMock
  - Add SetFallback in MockHttpRequestMiddleware. So others could intercept the mock request.
  - Update the previous luis mock approach as the fallback for HttpRequestMock
  - Update the previous luis mock approach to use HttpClientHandler instead of "luisResult"
  - Fix IncludeAPIResults not working in LuisAdaptiveRecognizer

## Testing

Add TestScriptTests_HttpRequestLuisMock to verify:
- previous approach is a fallback of HttpRequestMock
- content is matched partially
- IncludeAPIResults in LuisAdaptiveRecognizer

Update TestScriptTests_HttpRequestMock to verify fallback

### CODE REVIEW

File: tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/Tests/TestScriptTests/PropertyMock/propertiesluis.mock.dialog [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        {
+            "property": "settings.luis.endpointKey",
+            "value": "00000000000000000000000000000000"
+        },
+        {
+            "property": "settings.luis.endpoint",
+            "value": "https://westus.api.cognitive.microsoft.com"
+        }
+    ]
+}
```
Comment (chrimc62) [inline tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/Tests/TestScriptTests/PropertyMock/propertiesluis.mock.dialog:18]: This works, but it is a little different than the normal LUIS flow, i.e. you do luis:build and get a settings file.  I see you are caching the LUIS responses, but now I need to manually copy the settings into a file like this and run once to cache and once not--this makes it much more complicated to do the LUIS mocking.  I don't think you technically need to do that--can't you still use the regular properties file?

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/Schemas/Microsoft.Test.HttpRequestSequenceMock.schema [hunk 2]
```diff
@@ -32,6 +32,14 @@
...
         },
+        "content": {
+            "type": "string",
+            "title": "Content",
+            "description": "The content to match partially against.",
+            "examples": [
+                "content"
+            ]
+        },
         "responses": {
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/Schemas/Microsoft.Test.HttpRequestSequenceMock.schema:43]: I'm a little confused on how this is used.  Is this for matching against things like the post body?  If so, body would probably be a better name.

File: tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/Tests/TestScriptTests/HttpRequestMock/httpluis.mock.dialog [hunk 3]
```diff
@@ -0,0 +1,23 @@
+﻿{
+    "$schema": "../../../../tests.schema",
+    "$kind": "Microsoft.Test.HttpRequestSequenceMock",
+    "url": "https://westus.api.cognitive.microsoft.com/luis/prediction/v3.0/apps/00000000-0000-0000-0000-000000000000/*",
+    "content": "\"query\": \"overwrite\"",
```
Comment (chrimc62) [inline tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/Tests/TestScriptTests/HttpRequestMock/httpluis.mock.dialog:5]: You can use a pattern here that is not sensitive to the app id, right?

Comment (xieofxie) [reply tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/Tests/TestScriptTests/HttpRequestMock/httpluis.mock.dialog:5]: Yes. But I included an id here to give an example for how to match same query against different luis mocks (like top level and sub level would return different intents for same query)

File: tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/Tests/TestScriptTests/PropertyMock/propertiesluis.mock.dialog (hunk 1 above)
Comment (xieofxie) [reply tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/Tests/TestScriptTests/PropertyMock/propertiesluis.mock.dialog]: Updated.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/Schemas/Microsoft.Test.HttpRequestSequenceMock.schema (hunk 2 above)
Comment (xieofxie) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/Schemas/Microsoft.Test.HttpRequestSequenceMock.schema:57]: Add a new field BodyMatchType: Exact and Partial. I just use methods from RichardSzalay.MockHttp now. In the future, we could extend by adding regex etc.

File: tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/Tests/TestScriptTests/HttpRequestMock/httpluis.mock.dialog (hunk 3 above)
Comment (chrimc62) [reply tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Tests/Tests/TestScriptTests/HttpRequestMock/httpluis.mock.dialog]: But this kind of my complaint.  LUIS is complicated to mock which is why getting a real response from LUIS and verifying it was the approach I took.  The case where you mock directly is much less likely.

---
In reply to: [471885860](https://github.com/microsoft/botbuilder-dotnet/pull/4453#discussion_r471885860) [](ancestors = 471885860)

**Comment by xieofxie** [comment]
Oh, no, server is down https://schemas.botframework.com/schemas/component/v1.0/component.schema

**Comment by xieofxie** [comment]
Hi @mrivera-ms could you review and merge this? Thanks!

**Pull request merged**